### PR TITLE
Update ownerclaims for apply and non-apply cases

### PR DESCRIPTION
**Pull request by jennybuckley**

### CODE REVIEW

File: fieldpath/fromdiff.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+package fieldpath
+
+import (
+	"github.com/kubernetes-sigs/structured-merge-diff/value"
+)
+
+// SetFromDiff creates a set containing every leaf field that has a
+// different value between v and u.
+func SetFromDiff(v, u value.Value) *Set {
```
Comment (lavalamp) [inline fieldpath/fromdiff.go]: This doesn't need the schema?

File: merge/object.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+*/
+
+package merge
+
+import (
+	"github.com/kubernetes-sigs/structured-merge-diff/fieldpath"
+	"github.com/kubernetes-sigs/structured-merge-diff/value"
+)
+
+type Version string
```
Comment (lavalamp) [inline merge/object.go]: Godoc all the things.

File: merge/object.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	*fieldpath.Set
+	Version Version
+}
+
+type Converter interface {
+	Convert(object *Object, version Version) *Object
+}
+
+type Object struct {
+	value.Value
```
Comment (lavalamp) [inline merge/object.go]: Typed?

File: merge/object.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+func (o *Object) OwnerSet() map[string]*VersionedSet          { return nil }
+func (o *Object) SetOwnerSet(owners map[string]*VersionedSet) {}
+func (o *Object) GroupVersion() Version                       { return Version("v1") }
```
Comment (lavalamp) [inline merge/object.go]: I think the intention is that apiserver can construct one of these that works with kubernetes api objects?

Maybe use a fieldpath to locate the version field and another fieldpath to locate the workflow id map?

File: merge/object.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+func ConflictsToError(conflicts map[string]*VersionedSet) error { return nil }
```
Comment (lavalamp) [inline merge/object.go]: Define a type for conflicts? I don't understand where the map comes in.

File: merge/object.go (hunk 4 above)
Comment (lavalamp) [reply merge/object.go]: Or that stuff could go in the schema?

File: merge/object.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+func MergeApplied(live, config *Object) *Object                 { return nil }
```
Comment (lavalamp) [inline merge/object.go]: Should this take objects or typed.TypedValue?

File: merge/update.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	}
+
+	for owner, conflictSet := range conflicts {
+		owners[owner].Set = owners[owner].Set.Difference(conflictSet.Set)
+	}
+
+	return owners, nil
+}
+
+func (s *SetUpdater) NonApply(live, nw *Object, owner string) *Object {
```
Comment (lavalamp) [inline merge/update.go]: Maybe Update, Put, or Patch?

File: merge/update.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+limitations under the License.
+*/
+
+package merge
+
+import (
+	"github.com/kubernetes-sigs/structured-merge-diff/fieldpath"
+)
+
+type SetUpdater struct {
```
Comment (lavalamp) [inline merge/update.go]: Applier? WorkflowOperations?

Godoc :)

File: merge/update.go (hunk 7 above)
Comment (lavalamp) [reply merge/update.go]: Hm, there's already an Update. Godoc comments would help a lot :)

File: merge/update.go [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+func (s *SetUpdater) NonApply(live, nw *Object, owner string) *Object {
+	owners, _ := s.Update(live, nw, owner, true)
+	changed := fieldpath.SetFromDiff(live.Value, nw.Value)
+	removed := fieldpath.SetFromValue(live.Value).Difference(fieldpath.SetFromValue(nw.Value))
+	owners[owner].Set = owners[owner].Set.Union(changed).Difference(removed)
+	nw.SetOwnerSet(owners)
+	return nw
+}
+
+func (s *SetUpdater) Apply(live, config *Object, owner string, force bool) (*Object, error) {
```
Comment (lavalamp) [inline merge/update.go]: is owner the same as workflow? Maybe use the same name everywhere (I vote for workflow but don't care all that much).

File: fieldpath/fromdiff.go (hunk 1 above)
Comment (jennybuckley) [reply fieldpath/fromdiff.go]: It probably does need the schema eventually but this way it keeps consistency with SetFromValue.

File: merge/object.go (hunk 3 above)
Comment (jennybuckley) [reply merge/object.go]: I think the type can be inferred from the Value, if we look at the apiVersion field

Comment (jennybuckley) [reply merge/object.go]: But seeing your comment below I like the idea of this being implemented in the apiserver better

File: merge/object.go (hunk 5 above)
Comment (jennybuckley) [reply merge/object.go]: map from workflowid to fields that it owns which are conflicting with the apply request

Comment (jennybuckley) [reply merge/object.go]: So we can return the owner of the conflicting fields to the user

File: merge/update.go (hunk 9 above)
Comment (kwiesmueller) [reply merge/update.go]: Just asked this question to @apelisse as he was using owner in #25 
I could understand the favor for owner as it is easier to understand and does cover more cases (while workflow would sound weird for users or operators in the future.

**Comment by apelisse** [comment]
This has been merged through a different PR (#32)

**Pull request closed without merging**